### PR TITLE
feat(#1461): add stale worktree + DerivedData prune script

### LINKED ISSUE #1461: Add a maintenance script to prune stale worktrees + DerivedData
## Summary

`.claude/worktrees/` and `~/Library/Developer/Xcode/DerivedData` currently rely entirely on humans/agents remembering to `git worktree remove` after a PR merges (documented in `CLAUDE.md` ▸ "Worktrees", added in #). In practice this doesn't happen reliably: an audit on 2026-08-13 found 43 directories under `.claude/worktrees/` against only ~22 registered `git worktree list` entries, several tied to PRs that had already merged, totaling 141GB — plus a further 146GB in orphaned `DerivedData` folders (one per worktree's distinct `Anglesite.xcodeproj` path). One force-removal during that cleanup also collided with a still-running agent session, because git status and open-PR checks alone don't reveal a live process using a worktree.

## Proposal

A local maintenance script (not CI — this is host-machine state, not something CI touches) that can run periodically or on demand, e.g. `scripts/prune-stale-worktrees.sh`:

1. **Find orphaned directories** — diff `.claude/worktrees/*` on disk against `git worktree list --porcelain`; anything present on disk but not registered (broken `gitdir` links from a rename/relocation, or an interrupted removal) is a plain leftover directory.
2. **Find merged/closed candidates** — for each *registered* worktree, check its branch against `gh pr list --head <branch> --state all`; flag anything whose PR is merged or closed.
3. **Safety-check before touching anything** — for every candidate, require `git status --short` clean (or its PR already merged) **and** no live process holding it as a cwd (`lsof -a -d cwd` across `claude`/agent/build processes). This second check is the one that's easy to skip and caused the live-session collision above.
4. **Report by default, remove with a flag** — plain run prints candidates + reclaimable size; `--remove` (or similar) actually runs `git worktree remove` / `rm -rf` for orphaned dirs.
5. Optionally also report/clear stale `~/Library/Developer/Xcode/DerivedData/Anglesite-*` folders that no longer correspond to any current worktree path.

Open question: whether this should also be wired into an existing entry point (e.g. `scripts/setup-dev-env.sh` prompting when the count/size crosses a threshold) so it surfaces without anyone having to remember to run it, or stay a manual/on-demand tool. Worth deciding before implementation.

## Context

Follow-up from the worktree/DerivedData cleanup discussed in this session — see `CLAUDE.md` ▸ "Worktrees" for the manual procedure this would automate.

### PR DESCRIPTION
**Pull request by davidwkeith**
Closes #1461

## Summary

- `scripts/prune-worktrees.sh` automates the manual cleanup procedure from `AGENTS.md` ▸ "Worktrees": report-only by default; `--remove` deletes with per-candidate confirmation (`--yes` for non-interactive). Registered worktrees under `.claude/worktrees/` are candidates only when their branch's PR is **merged/closed** (`gh pr list --head … --state all`) *and* `git status --short` is clean — a dirty-but-merged worktree needs an explicit `--force-dirty` (never `git worktree remove --force` by default). Orphaned directories (on disk but not in `git worktree list` — the `comm -23` case) are `rm -rf`'d. **Every** candidate must first pass the lsof live-cwd check, since PR state and git status alone don't reveal a live agent session using a worktree (the collision that motivated #1461). `DerivedData/Anglesite-*` dirs whose `info.plist` `WorkspacePath` no longer exists — including ones freed by the same run's removals — are pruned too. Every keep/skip/candidate/removal is printed with its reason.
- `scripts/prune-worktrees.test.sh` (same convention as `install-swift-linux.test.sh`) exercises every safety branch against a throwaway repo with stubbed `gh`/`lsof`/`plutil`: merged+clean removed, merged+dirty gated on `--force-dirty`, open-PR kept, live-cwd skipped, orphan dir removed, stale vs. live DerivedData. Not wired into CI — per the issue, this is host-machine state CI doesn't have.
- `scripts/setup-dev-env.sh` (owner decision on #1461, 2026-08-15): when `.claude/worktrees/` crosses **>5 worktrees or ≥20GB**, prints a notice and — on an interactive terminal — offers to run the read-only prune report. Setup never deletes anything itself; removal always goes through the prune script's own `--remove` + safety checks. Resolves the main checkout via `--git-common-dir`, so both scripts work when run from inside a worktree.

## Paired PR check

- [x] This change is **self-contained** to `Anglesite/Anglesite`.
- [ ] This change **needs a paired PR** in [`Anglesite/anglesite-skills`](https://github.com/Anglesite/anglesite-skills) (MCP sidecar server). Link it here:

> Cross-cutting work (extending MCP messages) lands as paired PRs. The sidecar PR ships first in a tagged release; the app PR consumes it and re-vendors the container image. Template changes are app-only (`Resources/Template/`). See `CLAUDE.md` ▸ "Two-repo coordination".

## Test plan

- [x] `scripts/prune-worktrees.test.sh` — 26 assertions across report-only, `--remove --yes`, and `--force-dirty` runs, all passing (macOS; stubs keep it Linux-portable).
- [x] Manual: `scripts/prune-worktrees.sh` (report mode) against the real repo — correctly kept open/no-PR worktrees, skipped a merged worktree held live by an active agent session (naming pids), flagged 5 candidates (16.7GB), deleted nothing.
- [x] Manual: `scripts/setup-dev-env.sh` non-interactive run — printed the over-threshold notice (15 dirs, ~46GB) with the report/remove commands, skipped the prompt, exited 0.
- [ ] `swift test --package-path .` / `xcodebuild` — not run: scripts-only change, no Swift/JS touched.

### CODE REVIEW

**Review by davidwkeith — COMMENTED** [review]
Reviewed `prune-worktrees.sh` and the `setup-dev-env.sh` hygiene hook. Solid work — the safety checks (live-cwd guard via `lsof`, merged/dirty gating, orphan-directory detection, DerivedData staleness check) are all correct against the accompanying test suite, and report-only-by-default with per-candidate confirmation is the right posture for something this destructive.

One robustness nit, not blocking: `pr_state()` calls `gh pr list --head "$branch" ...` without pinning the repo (no `-R`/cwd anchor). It works today because the script is always invoked from inside this repo (main checkout or a worktree, both sharing `origin`), so `gh` can infer the repo from cwd. But if it's ever invoked from an unrelated cwd (e.g. a cron/launchd job), `gh` would fail to resolve the repo and every branch would silently fall back to `UNKNOWN` — a safe failure mode (nothing gets pruned) but a silent no-op that could be confusing. Consider anchoring it, e.g. `gh pr list -R "$(git -C "$MAIN_ROOT" config --get remote.origin.url)" ...` or running the `gh` call with cwd pinned to `$MAIN_ROOT`.

No functional bugs found otherwise.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/prune-worktrees.sh [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    fi
+    case "$out" in
+        *MERGED*) echo MERGED ;;
+        *CLOSED*) echo CLOSED ;;
+        *OPEN*)   echo OPEN ;;
+        *)        echo NONE ;;
+    esac
+}
+
+size_kb() { du -sk "$1" 2>/dev/null | awk '{ print $1 }'; }
```
Comment (davidwkeith) [inline scripts/prune-worktrees.sh:115]: **[correctness] `size_kb()`'s unguarded `du` pipeline can silently abort the whole script mid-run under `set -euo pipefail` (line 38).**

`size_kb() { du -sk "$1" 2>/dev/null | awk '{ print $1 }'; }` — with `pipefail`, if `du` fails (nonexistent/vanished dir, permission-denied subdir, broken symlink), the pipeline's exit status is non-zero even though `awk` itself succeeds. Every call site (`kb=$(size_kb "$wt")` at lines 190, 200, 235, 279) is a bare top-level assignment — not inside `if`/`&&`/`||` — so under `set -e` that non-zero status kills the entire script immediately, with no message (stderr is redirected to `/dev/null`).

Reproduced directly:
```
$ bash -c 'set -euo pipefail
size_kb() { du -sk "$1" 2>/dev/null | awk "{ print \$1 }"; }
kb=$(size_kb "/does/not/exist")
echo "reached end, kb=$kb"'
# → exits 1, "reached end" never prints
```

Given this repo runs many agents concurrently in `.claude/worktrees/` (per CLAUDE.md), a worktree/DerivedData dir vanishing or having permission quirks mid-scan is realistic — and since this can happen after Phase 1 has already `git worktree remove`d some candidates, the script would stop partway through with no summary and no explanation. Contrast with the `ws=$(plutil ... || true)` a few lines below (and the pre-existing `LIBXML2_NEW=$(... || true)` pattern in `setup-dev-env.sh`, whose comment explicitly calls out this exact `set -e` failure mode) — `size_kb()` is missing the same guard.

**Fix:** `size_kb() { du -sk "$1" 2>/dev/null | awk '{ print $1 }' || true; }` (or guard each call site).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/setup-dev-env.sh [hunk 2]
```diff
@@ -151,6 +151,45 @@ setup_linux() {
...
+check_worktree_hygiene() {
+    # Resolve the MAIN checkout — when setup runs from inside a worktree, the
+    # worktrees root lives beside the main checkout's .git, not this one.
+    local main_root wt_dir count size_kb size_gb
+    main_root=$(dirname "$(git -C "$REPO_ROOT" rev-parse --path-format=absolute --git-common-dir)")
+    wt_dir="$main_root/.claude/worktrees"
+    [[ -d "$wt_dir" ]] || return 0
+    count=$(find "$wt_dir" -mindepth 1 -maxdepth 1 -type d | wc -l | tr -d ' ')
+    [[ "$count" -gt 0 ]] || return 0
+    size_kb=$(du -sk "$wt_dir" 2>/dev/null | awk '{ print $1 }')
```
Comment (davidwkeith) [inline scripts/setup-dev-env.sh:170]: **[correctness] Same unguarded-`du` / `set -e` abort risk as `prune-worktrees.sh`'s `size_kb()`.**

`size_kb=$(du -sk "$wt_dir" 2>/dev/null | awk '{ print $1 }')` — if `du` fails on `$wt_dir` (e.g. a subdirectory removed by a concurrent agent between the `[[ -d "$wt_dir" ]]` check and this call, or a permission/broken-symlink issue somewhere under a multi-GB worktree tree), `pipefail` propagates the non-zero status and `set -e` (line 20) aborts the entire `setup-dev-env.sh` run right here — silently (stderr suppressed), before it ever gets to print the "Environment ready."/blocking-item summary. That defeats the "check-and-fix … prints actionable instructions" contract this script advertises for an informational hygiene nudge that's explicitly supposed to be best-effort. Notably, the pre-existing `LIBXML2_NEW=$(... || true)` a few dozen lines up in this same file has a comment explaining exactly this failure mode — this new call is missing that `|| true`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/prune-worktrees.sh [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+MAIN_ROOT=$(dirname "$(git -C "$SCRIPT_DIR" rev-parse --path-format=absolute --git-common-dir)")
+WORKTREES_DIR="$MAIN_ROOT/.claude/worktrees"
+DERIVED_DATA_ROOT="$HOME/Library/Developer/Xcode/DerivedData"
+
+# One lsof pass, cached: "cwd<TAB>pid<TAB>command" per live process. Used to
+# refuse any candidate that is some process's current working directory.
+LIVE_CWDS=$(lsof -d cwd -F pcn 2>/dev/null | awk '
+    /^p/ { pid = substr($0, 2) }
+    /^c/ { cmd = substr($0, 2) }
+    /^n/ { printf "%s\t%s\t%s\n", substr($0, 2), pid, cmd }' || true)
```
Comment (davidwkeith) [inline scripts/prune-worktrees.sh:82]: **[correctness] The live-cwd safety check is a single snapshot taken once at startup, never refreshed before an actual deletion — a TOCTOU gap in exactly the protection this script exists to provide.**

`LIVE_CWDS` is captured once here via one `lsof` pass, and `live_users()` (called per-candidate in Phases 1–2) only ever reads this same static snapshot. Between the snapshot and the actual `git worktree remove`/`rm -rf` for a given candidate, the script does a `du -sk` scan (can be slow on large worktrees — the PR's own test plan mentions "5 candidates (16.7GB)") and, without `--yes`, waits on an interactive `read -p` confirmation that can be open indefinitely.

Given this repo's own worktree convention is "multiple agents run in parallel" (CLAUDE.md), a process legitimately starting to use a merged-PR worktree as its cwd *after* the lsof snapshot but *before* the confirm/remove for that candidate is exactly the "collision that motivated #1461" this script's header describes — except now the tool built to prevent it can itself race past a session that starts up mid-run, because it never re-checks liveness at the moment of deletion.

**Fix:** re-run the lsof check (or at least re-check `live_users` against a fresh snapshot) immediately before each `git worktree remove` / `rm -rf`, not just once at the top.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/setup-dev-env.sh [hunk 4]
```diff
@@ -151,6 +151,45 @@ setup_linux() {
...
+    size_gb=$(( ${size_kb:-0} / 1048576 ))
```
Comment (davidwkeith) [inline scripts/setup-dev-env.sh:171]: **[correctness, minor] Integer-division truncation can suppress the hygiene notice right at the threshold boundary.**

`size_gb=$(( ${size_kb:-0} / 1048576 ))` truncates toward zero, so e.g. 19.99GB of worktrees computes `size_gb=19`, and the check at line 174 (`"$size_gb" -lt "$WORKTREE_SIZE_GB_THRESHOLD"` with threshold 20) reports "ok" and never surfaces the notice even though the tree is effectively at the 20GB line. Comparing against the un-truncated KB value (`size_kb -lt $((WORKTREE_SIZE_GB_THRESHOLD * 1048576))`) would avoid the rounding gap.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by davidwkeith — COMMENTED** [review]
Automated code review (high effort, single-pass — the Agent tool wasn't available in this context, so this is one pass through the diff rather than the usual multi-agent fan-out + verify pass). 4 findings posted inline, ranked most severe first. Verified two of them (the `set -e`/pipefail abort and the stale-snapshot TOCTOU) empirically against the actual shell semantics before posting; the diff's own test suite (`prune-worktrees.test.sh`, 26 assertions) was also run locally and passes as claimed — these findings are edge cases the tests don't cover.

---
_Generated by [Claude Code](https://claude.ai/code)_
